File: apps/ios/TesseroneKit/Sources/TesseroneKit/Barcode.swift
import Foundation

/// Port of `src/services/scanner.ts`.
public enum Barcode {
    public static let formatOptions: [BarcodeFormatOption] = [
        .init(value: .ean13, label: "EAN-13"),
        .init(value: .ean8, label: "EAN-8"),
        .init(value: .code128, label: "Code 128"),
        .init(value: .code39, label: "Code 39"),
        .init(value: .qr, label: "QR"),
        .init(value: .upcA, label: "UPC-A"),
        .init(value: .upcE, label: "UPC-E"),
    ]

    public static func mapScannerType(_ type: String) -> BarcodeFormat {
        BarcodeFormat.normalize(type)
    }

    public static func validate(_ code: String, format: BarcodeFormat) -> Bool {
        let trimmed = code.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return false }
        switch format {
        case .ean13:
            return trimmed.range(of: #"^\d{13}$"#, options: .regularExpression) != nil
        case .ean8:
            return trimmed.range(of: #"^\d{8}$"#, options: .regularExpression) != nil
        case .upcA:
            return trimmed.range(of: #"^\d{12}$"#, options: .regularExpression) != nil
        case .upcE:
            return trimmed.range(of: #"^\d{6,8}$"#, options: .regularExpression) != nil
        case .code128, .code39:
            return trimmed.count > 0
        default:
            return true
        }
    }

    /// EAN-13 sometimes scans as 12 digits (missing leading zero).
    public static func fixScannedCode(_ code: String, format: BarcodeFormat) -> (code: String, format: BarcodeFormat) {
        if format == .ean13, code.range(of: #"^\d{12}$"#, options: .regularExpression) != nil {
            return ("0" + code, .ean13)
        }
        return (code, format)
    }
}

File: apps/android/core/src/main/java/com/chipcolate/tesserone/core/barcode/Barcode.kt
package com.chipcolate.tesserone.core.barcode

import com.chipcolate.tesserone.core.model.BarcodeFormat

data class BarcodeFormatOption(
    val value: BarcodeFormat,
    val label: String,
)

/** Selectable formats for add/edit. Order is UI order. */
val BARCODE_FORMAT_OPTIONS: List<BarcodeFormatOption> = listOf(
    BarcodeFormatOption(BarcodeFormat.EAN13, "EAN-13"),
    BarcodeFormatOption(BarcodeFormat.EAN8, "EAN-8"),
    BarcodeFormatOption(BarcodeFormat.CODE128, "Code 128"),
    BarcodeFormatOption(BarcodeFormat.CODE39, "Code 39"),
    BarcodeFormatOption(BarcodeFormat.QR, "QR"),
    BarcodeFormatOption(BarcodeFormat.UPCA, "UPC-A"),
    BarcodeFormatOption(BarcodeFormat.UPCE, "UPC-E"),
)

fun mapBarcodeType(type: String): BarcodeFormat {
    val map = mapOf(
        "qr" to BarcodeFormat.QR,
        "ean13" to BarcodeFormat.EAN13,
        "ean8" to BarcodeFormat.EAN8,
        "code128" to BarcodeFormat.CODE128,
        "code39" to BarcodeFormat.CODE39,
        "upc_a" to BarcodeFormat.UPCA,
        "upc_e" to BarcodeFormat.UPCE,
        "pdf417" to BarcodeFormat.PDF417,
        "aztec" to BarcodeFormat.AZTEC,
        "datamatrix" to BarcodeFormat.DATAMATRIX,
        "itf14" to BarcodeFormat.ITF14,
        "itf" to BarcodeFormat.ITF14,
    )
    return map[type.lowercase()] ?: BarcodeFormat.CODE128
}

/**
 * Formats the Apple Watch companion renders natively. AZTEC / PDF417 / UPC-E /
 * DATAMATRIX fall back to "Open on phone" on both watch surfaces.
 */
fun BarcodeFormat.isWatchRenderable(): Boolean = when (this) {
    BarcodeFormat.QR,
    BarcodeFormat.EAN13,
    BarcodeFormat.EAN8,
    BarcodeFormat.CODE128,
    BarcodeFormat.CODE39,
    BarcodeFormat.UPCA,
    BarcodeFormat.ITF14,
    -> true
    BarcodeFormat.UPCE,
    BarcodeFormat.PDF417,
    BarcodeFormat.AZTEC,
    BarcodeFormat.DATAMATRIX,
    -> false
}

fun validateBarcode(code: String, format: BarcodeFormat): Boolean {
    val c = code.trim()
    if (c.isEmpty()) return false
    return when (format) {
        BarcodeFormat.EAN13 -> Regex("""^\d{13}$""").matches(c)
        BarcodeFormat.EAN8 -> Regex("""^\d{8}$""").matches(c)
        BarcodeFormat.UPCA -> Regex("""^\d{12}$""").matches(c)
        BarcodeFormat.UPCE -> Regex("""^\d{6,8}$""").matches(c)
        BarcodeFormat.CODE128, BarcodeFormat.CODE39 -> c.isNotEmpty()
        else -> true
    }
}

data class ScannedCode(
    val code: String,
    val format: BarcodeFormat,
)

/** EAN13 scans sometimes drop the leading zero (12 digits). */
fun fixScannedCode(code: String, format: BarcodeFormat): ScannedCode {
    if (format == BarcodeFormat.EAN13 && Regex("""^\d{12}$""").matches(code)) {
        return ScannedCode("0$code", BarcodeFormat.EAN13)
    }
    return ScannedCode(code, format)
}
